opensearch/src/main/java/org/opensearch/sql/opensearch/storage/script/filter/ExpressionFilterScript.java

```diff
@@ -48,6 +48,15 @@ private ExprValue evaluateExpression(
       return ExprBooleanValue.of(false);
     }
 
+    // refer to https://github.com/opensearch-project/sql/issues/2796
+    if (result.type() == ExprCoreType.INTEGER) {
+      if (result.integerValue() == 0) {
+        result = ExprBooleanValue.of(false);
+      } else if (result.integerValue() == 1) {
+        result = ExprBooleanValue.of(true);
```

Comment:
This check will cause additional issue.
`select * from %s where 1 * 1` will be treated as `where true` and `select * from %s where 1 * 0` will be treated as `where false`. They should throw exceptions.
How about change the return value type to `ExprBooleanValue` in https://github.com/opensearch-project/sql/blob/43263963aca2ada949d55ae01cd38788d5d96fbb/core/src/main/java/org/opensearch/sql/utils/OperatorUtils.java#L37-L40

Reply:
If we modify the return type of OperatorUtils#matchesRegexp, it will impact the return type of the REGEXP function, which could be a breaking change. For instance, the query `SELECT 'Hello!' REGEXP '.*', 'a' REGEXP 'b'` will initially yield `1,0` as the output. However, after modifying the return type, the output will be `true,false`.

Reply:
@LantaoJin only when the expression is a REGEXP expression will the conversion be executed. What do you think?

Reply:
Sounds good. The conversion scope should be limited in `REGEXP` expression.

Reply:
I have fixed it. Please take another look at your convenience.

Reply:
Could you explain why `REGEXP` should return integer? From the Javadoc, the implementation seems wrong? I'm thinking we should avoid adding such special logic on critical path if possible.

```
  /**
   * Checks if text matches regular expression pattern.
   *
   * @param pattern string pattern to match.
   * @return if text matches pattern returns true; else return false.
   */
  public static ExprIntegerValue matchesRegexp(ExprValue text, ExprValue pattern) {
    return new ExprIntegerValue(
        Pattern.compile(pattern.stringValue()).matcher(text.stringValue()).matches() ? 1 : 0);
  }
```

Reply:
> Could you explain why REGEXP should return integer?

@dai-chen I guess the reason is it's the behaviour of MySQL (https://dev.mysql.com/doc/refman/8.0/en/regexp.html#operator_regexp). The original request was opened in https://github.com/opendistro-for-elasticsearch/sql/issues/710 and its implementation introduced by https://github.com/opendistro-for-elasticsearch/sql/pull/750. It would be a breaking change if we change its return value type. Besides updating the user doc (https://github.com/opensearch-project/sql/blob/main/docs/user/dql/expressions.rst#regexp-value-test), we need a changelog file to record those breaking changes. cc @penghuo, @chloe-zh

Reply:
My proposal are two options:
1. Correct the return type of `matchesRegexp()` to `ExprBooleanValue` in this PR, including user doc updates, then we need a new PR to introduce a changelog to record any breaking changes.
2. Close current RP because it is not an issue. `SELECT field1 FROM test WHERE 1 = (field1 REGEXP 'test.*')` should work. The current fixing not only adds special logic but also introduces new confused behaviour (REGEXP expression returns integer in SELECT clause but boolean in WHERE clause).

Reply:
In my opinion, option 2 lacks user-friendliness.